src/tests/neighborhood_api.test.ts

```diff
@@ -38,3 +38,116 @@ describe('When no neighborhood exists in the db', () => {
   });
 });
+
+describe('Testing UPDATE method for neighborhood API.', () => {
+  beforeEach(async () => {
+    await seed();
+  });
+
+  test('Update all of a neighborhood\'s fields by id', async () => {
+    const neighborhoodToUpdate = await prismaClient.neighborhood.findFirst({
+      where: {
+        name: "Antonina's Neighborhood",
+      },
+    });
+
+    const newAdmin = await prismaClient.user.findFirst({
+      where: {
+        NOT: { id: neighborhoodToUpdate!.admin_id },
+      },
+    });
+
+    const newData = {
+      name: 'Test',
+      description: 'Test',
+      admin_id: newAdmin!.id,
+      location: 'Athens',
+    };
+
+    const response = await api.put(`/api/neighborhoods/${neighborhoodToUpdate!.id}`).send(newData);
+
+    expect(response.text).toEqual('Neighborhood \'Test\' has been updated.');
+    expect(response.status).toEqual(200);
+    expect(await prismaClient.neighborhood.findFirst({
+      where: { id: neighborhoodToUpdate!.id },
+    })).toEqual({
+      id: neighborhoodToUpdate!.id,
+      ...newData,
+    });
+  });
+
+  test('Partial update works', async () => {
+    const neighborhoodToUpdate = await prismaClient.neighborhood.findFirst({
+      where: {
+        name: "Antonina's Neighborhood",
+      },
+    });
+
+    const newData = { name: 'Test' };
+    const response = await api.put(`/api/neighborhoods/${neighborhoodToUpdate!.id}`).send(newData);
+
+    expect(response.text).toEqual('Neighborhood \'Test\' has been updated.');
+    expect(response.status).toEqual(200);
+    expect(await prismaClient.neighborhood.findFirst({
+      where: { id: neighborhoodToUpdate!.id },
+    })).toEqual({
+      ...neighborhoodToUpdate,
+      name: newData.name,
+    });
+  });
+
+  test('Empty input doesn\'t change anything on the server', async () => {
+    const neighborhoodToUpdate = await prismaClient.neighborhood.findFirst({
+      where: {
+        name: "Antonina's Neighborhood",
+      },
+    });
+
+    const response = await api.put(`/api/neighborhoods/${neighborhoodToUpdate!.id}`).send({});
+
+    expect(response.text).toEqual('Neighborhood \'Antonina\'s Neighborhood\' has been updated.');
+    expect(response.status).toEqual(200);
+    expect(await prismaClient.neighborhood.findFirst({
+      where: { id: neighborhoodToUpdate!.id },
+    })).toEqual({
+      ...neighborhoodToUpdate,
+    });
+  });
+
+  test('Update with invalid properties raises an error', async () => {
+    const neighborhoodToUpdate = await prismaClient.neighborhood.findFirst({
+      where: {
+        name: "Antonina's Neighborhood",
+      },
+    });
+
+    const newData = {
+      name: 'Test',
+      description: 'Test',
+      location: 'Athens',
+      invalid: 'Non-existent prop',
+    };
+
+    const response = await api.put(`/api/neighborhoods/${neighborhoodToUpdate!.id}`).send(newData);
+    expect(response.status).toEqual(400);
+  });
+
+  test('Non-existent admin_id raises a 404 error', async () => {
```

Comment:
Looks good, Antonina. Very thorough! Radu, I don't think that that can come up in the real world, unless someone directly hits the route from the browser. My thinking is that a user will see a list of neighborhoods and either view or edit one of them. (Come to think of it, I'm not sure that the test I just wrote is possible.) Any such failures are application failures, since users will be selecting from valid lists.

I'm thinking now that if we want to change the admin for a neighborhood, we'll be assigning the new admin from a list of existing users, right? Like a select box. So, if there's an invalid user id entered that should be an abend (abnormal end, "contact your system admin" kind of thing).

Reply:
You're right, it doesn't make sense to check for that editing a non-existent neighborhood here, because only the owner can edit it. But in your branch it does make sense; another user could try and access a neighborhood that has been recently deleted and the server didn't update fast enough or another user has sent a link to a deleted neighborhood.

Reply:
That's actually a good catch because as the error handler was so far, if a different kind of Prisma error occurred, such as the neighborhood wasn't found, the program was hanging (the error wasn't handled in any way in the middleware). Hopefully I've fixed that in the latest commits.